hi-ml/src/health_ml/utils/checkpoint_handler.py

```diff
@@ -0,0 +1,139 @@
+from typing import List, Optional
+from urllib.parse import urlparse
+
+import requests
+from azureml.core import Run
+
+from health_azure.utils import is_global_rank_zero
+from health_ml.lightning_container import LightningContainer
+from health_ml.utils.checkpoint_utils import (MODEL_WEIGHTS_DIR_NAME, find_recovery_checkpoint_on_disk_or_cloud)
+
+
+class CheckpointHandler:
+    """
+    This class handles which checkpoints are used to initialize the model during train or test time
+    """
+
+    def __init__(self,
+                 container: LightningContainer,
+                 project_root: Path,
+                 run_context: Optional[Run] = None):
+        self.container = container
+        self.project_root = project_root
+        self.run_context = run_context
+        self.trained_weights_paths: List[Path] = []
+        self.has_continued_training = False
+
+    def download_recovery_checkpoints_or_weights(self) -> None:
+        """
+        Download checkpoints from a run recovery object or from a weights url. Set the checkpoints path based on the
+        run_recovery_object, weights_url or local_weights_path.
+        This is called at the start of training.
+
+        :param: only_return_path: if True, return a RunRecovery object with the path to the checkpoint without actually
+        downloading the checkpoints. This is useful to avoid duplicating checkpoint download when running on multiple
+        nodes. If False, return the RunRecovery object and download the checkpoint to disk.
+        """
+        if self.container.weights_url or self.container.local_weights_path:
+            self.trained_weights_paths = self.get_local_checkpoints_path_or_download()
+
+    def additional_training_done(self) -> None:
+        """
+        Lets the handler know that training was done in this run.
+        """
+        self.has_continued_training = True
+
+    def get_recovery_or_checkpoint_path_train(self) -> Optional[Path]:
+        """
+        Decides the checkpoint path to use for the current training run. Looks for the latest checkpoint in the
+        checkpoint folder. If run_recovery is provided, the checkpoints will have been downloaded to this folder
+        prior to calling this function. Else, if the run gets pre-empted and automatically restarted in AML,
+        the latest checkpoint will be present in this folder too.
+
+        :return: Constructed checkpoint path to recover from.
+        """
+        if is_global_rank_zero():
+            checkpoints = list(self.container.checkpoint_folder.rglob("*"))
+            logging.info(f"Available checkpoints: {len(checkpoints)}")
+            for f in checkpoints:
+                logging.info(f)
+        return find_recovery_checkpoint_on_disk_or_cloud(self.container.checkpoint_folder)
+
+    def get_checkpoint_to_test(self) -> Path:
+        """
+        Find the model checkpoint that should be used for inference. If a run recovery is provided, or if the model
```

Comment:
Checking if the run recovery case is included somewhere (earlier we had a case with recovery checkpoints `if self.run_recovery or self.has_continued_training`) If this isn't required now, perhaps the docs need to be updated?

Reply:
Thanks, updated.